test/runnable/test19065.d

```diff
@@ -0,0 +1,29 @@
+module test19065;
+
+struct S
+{
+    bool b = false;
+
+    invariant
+    {
+        assert(b == true);
+    }
+
+    @disable this();
+    @safe this(int)
+    {
+        b = true;
+    }
+
+    // Invariant may be violated in the destructor.
+    // This is because `this` may also be S.init.
+    // For this reason, the invariant must not be checked on struct entry.
+    @safe ~this()
+    {
+    }
+}
+
+@safe void main()
+{
+    S s = S.init;
+}
```

Comment:
Will the opposite fail? i.e.
```
struct S {
  bool value = false;
  invariant {
    assert(value == false);
  }
  @disable this();
}

void main() {
  S s = S.init;
  s.value = true;
}
```

I ask because I wonder if the check to see if the invariant needs to be called before the destructor depends on a by value comparison of `this` and `T.init` or an address?

If it's a by value comparison then will it work for classes where `this == other` compares addresses?

Reply:
There is no such check. That was something that'd been proposed, but I don't know how to write it.

I'm going to close this PR. It doesn't do anything good or worthwhile, and I found a better way. (Unions skip destructors, apparently _intentionally??_)